File: algo_problems/q541-560/q547.py
class Solution:
    def findCircleNum(self, M):
        """
        :type M: List[List[int]]
        :rtype: int
        """
        n = len(M)
        if n == 1:
            return 1
        found = [False for i in range(n)]
        result = 0
        stack = []
        for i in range(n):
            if found[i] is True:
                continue
            result += 1
            stack.append(i)
            while len(stack) != 0:
                j = stack.pop()
                found[j] = True
                for k in range(n):
                    if found[k] is not True and M[j][k] == 1:
                        found[k] = True
                        stack.append(k)
        return result


if __name__ == '__main__':
    print(Solution().findCircleNum(
        [
            [1, 1, 0],
            [1, 1, 0],
            [0, 0, 1]
        ]
    ))
    print(Solution().findCircleNum(
        [[1, 0, 0, 1],
         [0, 1, 1, 0],
         [0, 1, 1, 1],
         [1, 0, 1, 1]]
    ))

File: algo_problems/q541-560/q547.java
public class Solution {
    public int findCircleNum(int[][] M) {
    	if (M.length == 1){
    		return 1;
    	}
    	boolean[] visited = new boolean[M.length];
    	for (int i = 0; i < M.length; i ++) {
    		visited[i] = false;
    	}
    	Stack <Integer> stack = new Stack();
    	int result = 0;
    	for (int i = 0; i < M.length; i ++) {
    		if (visited[i] == true) {
    			continue;
    		}
    		result += 1;
    		stack.push(i);
    		visited[i] = true;
    		while(stack.size() != 0) {
    			Integer t = stack.pop();
    			for (int j = 0; j < M.length; j ++) {
    				if (visited[j] == false && M[t][j] == 1 ) {
    					visited[j] = true;
    					stack.push(j);
    				}
    			}
    		}
    	}
    	return result;

    }
}